File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.cpp
class SmallestInfiniteSet {
public:
    priority_queue<int,vector<int>, greater<int>> minq; 
    set<int> available; 
    SmallestInfiniteSet() {
        for(int i = 1; i <= 1000; i++){
            minq.push(i); 
        }
    }
    
    int popSmallest() {
        int res = minq.top();
        available.insert(res);
        minq.pop(); 
        return res; 
    }
    
    void addBack(int num) {
        if(available.contains(num)){
            available.erase(num); 
            minq.push(num); 
        }
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.java
class SmallestInfiniteSet {

    PriorityQueue<Integer> minHeap; 
    Set<Integer> hashSet; 
    int nextInt = 1; 

    public SmallestInfiniteSet() {
        minHeap = new PriorityQueue<>(); 
        hashSet = new HashSet<>(); 
    }
    
    public int popSmallest() {
        if(!minHeap.isEmpty()){
            int min_ = minHeap.poll(); 
            if(hashSet.contains(min_)) hashSet.remove(min_); 
            return min_; 
        }
        return nextInt++; 
    }
    
    public void addBack(int num) {
        if(nextInt > num && !hashSet.contains(num)){
            hashSet.add(num); 
            minHeap.offer(num); 
        }
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet obj = new SmallestInfiniteSet();
 * int param_1 = obj.popSmallest();
 * obj.addBack(num);
 */